### PR TITLE
Targetable emotes bring up a list of things to target

### PR DESCRIPTION
**Pull request by Superlagg**
[FEATURE]
## About the PR 

Certain targetable emotes now bring up a list of things to interact with if not given a target. For instance, you do "say *hug" and it'll display a list of nearby objs and mobs you can hug. If the thing moves out of interaction range (for things where distance matters), it'll return a message about how the thing isn't in range.

Emotes that do this!
- *hug, *sidehug (Hugs passes the emote params, which suppresses the dialog box)
- *blowkiss
- *bow
- *nodto (*nod doesn't do this, cus keybinds)
- *saluteto (ditto)
- *waveto (same)
- *glare, *look, *stare, *leer
- *flipoff, *flipbird, *middlefinger (diluted fliptonium doesn't trigger these) (Also ignores distance checks)
- *doubleflip, *doubledeuce, *doublebird, *flip2 (Ignores distance checks too)
- *boggle (Ignores distance check)
- *shakefist (Ignores distance check)
- *handshake, *shakehand, *shakehands (Mobs only)
- *dap, *daps (the heck is a dap) (Mobs only)
- *slap, *bitchslap, *smack (Mobs only)
- *highfive (Mobs only)

Also splits *salute, *wave, and *nod each into two emotes, one using the target dialog, and the other not. The ones that use the target dialog are *saluteto, *waveto, and *nodto, so these can be used with this without making the hotkey emotes annoying to spam.

Also, most of these emotes also accept objs as their target, so you can blow kisses to vending machines and shake their fists at Beepsky.

## Why's this needed? 

Its always been soo inconvenient to hug people. I like to flip people off, but don't like typing out their name.

## Changelog

```changelog
(u)Superlagg
(+)Most targetable emotes now bring up a menu of valid targets.
```

### CODE REVIEW

**Comment by Adharainspace** [comment]
what if we datumised emotes instead ha ha ...

File: code/mob/living/carbon/human/procs/emote.dm [hunk 1]
```diff
@@ -2215,3 +2331,21 @@
...
+	. = list()
+
+	if(kind_of_target == "mob" || kind_of_target == "both")
+		for(var/mob/M in view(range, get_turf(src)))
+			if(M == src)
+				continue
+			. |= M
+
+	if(kind_of_target == "obj" || kind_of_target == "both")
+		for(var/obj/O in view(range, get_turf(src)))
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:2348]: could shove for (for atom movable in view) into a varlist and then depending on target get just the mobs or objects from the list

File: code/mob/living/carbon/human/procs/emote.dm [hunk 2]
```diff
@@ -1182,11 +1290,19 @@
...
 									break
-						if (M)
 #ifdef TWITCH_BOT_ALLOWED
 							if (IS_TWITCH_CONTROLLED(M))
 								return
 #endif
+						else
+							var/list/target_list = src.get_targets(1, "mob") // Chrunb Erbrbt and Scales To Lizard highfive!
+							if(length(target_list))
+								M = input(src, "Pick someone to high-five!", "EmotiConsole v1.1.3") as null|anything in target_list
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:1300]: i'd like to see this using the tgui input list

File: code/mob/living/carbon/human/procs/emote.dm [hunk 3]
```diff
@@ -1182,11 +1290,19 @@
...
+								if (M && !IN_RANGE(get_turf(src), get_turf(M), 1))
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:1301]: you probably want interact dist (goes for all IN_RANGE checks)

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/mob/living/carbon/human/procs/emote.dm (hunk 1 above)
Comment (Superlagg) [reply code/mob/living/carbon/human/procs/emote.dm]: I think I did that, though I'm not really sure what you mean by that.

File: code/mob/living/carbon/human/procs/emote.dm (hunk 2 above)
Comment (Superlagg) [reply code/mob/living/carbon/human/procs/emote.dm]: Aight, made them use the tgui input list, and it looks like it works pretty good!

File: code/mob/living/carbon/human/procs/emote.dm (hunk 3 above)
Comment (Superlagg) [reply code/mob/living/carbon/human/procs/emote.dm]: in_interact_range looks like it only checks the tiles right next to the player, so I made the ones that require physical contact use it, and left the ranged ones to IN_RANGE.

File: code/mob/living/carbon/human/procs/emote.dm [hunk 4]
```diff
@@ -2215,3 +2340,30 @@
 					sleep(1 SECOND) //let us do that combo shit people like with throwing
 					src.force_laydown_standup()
 
+/// Looks for the kind_of_target atoms within range, and throws the user an input
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:2343]:
```suggestion
/// Looks for the kind_of_target movables within range, and throws the user an input
```

File: code/mob/living/carbon/human/procs/emote.dm [hunk 5]
```diff
@@ -1182,11 +1299,19 @@
...
 									break
-						if (M)
 #ifdef TWITCH_BOT_ALLOWED
 							if (IS_TWITCH_CONTROLLED(M))
 								return
 #endif
+						else
+							var/list/target_list = src.get_targets(1, "mob") // Chrunb Erbrbt and Scales To Lizard highfive!
+							if(length(target_list))
+								M = tgui_input_list(src, "Pick someone to high-five!", "EmotiConsole v1.1.3", target_list, 0)
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:1309]: i think having a timeout would be good for these, even if it's a whole minute

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: code/mob/living/carbon/human/procs/emote.dm (hunk 5 above)
Comment (Superlagg) [reply code/mob/living/carbon/human/procs/emote.dm]: Okay, set there to be a timeout of about 20 seconds.

File: code/mob/living/carbon/human/procs/emote.dm [hunk 6]
```diff
@@ -2234,3 +2357,31 @@
 					G.affecting.force_laydown_standup()
 					sleep(1 SECOND) //let us do that combo shit people like with throwing
 					src.force_laydown_standup()
+
+/// Looks for the kind_of_target movables within range, and throws the user an input
+/// Valid kinds: "mob", "obj", "both"
+/mob/living/proc/get_targets(range = 1, kind_of_target = "mob")
+	if(!isturf(get_turf(src))) return
+
+	var/list/everything_around = list()
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:2366]:
```suggestion
	var/list/atom/movable/everything_around = list()
```

File: code/mob/living/carbon/human/procs/emote.dm [hunk 7]
```diff
@@ -334,32 +366,51 @@
...
+							boutput(src, "<span class='emote'><B>[M]</B> is [inaction_phrase]!</span>")
+							return
 
 				act = lowertext(act)
-				if (param)
+				if (M)
 					switch(act)
 						if ("nod")
 							message = "<B>[src]</B> [act]s to [param]."
 							maptext_out = "<I>[act]s to [param]</I>"
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:409]:
```suggestion
						if ("nod")
							message = "<B>[src]</B> [act]s to [M]."
							maptext_out = "<I>[act]s to [M]</I>"
```

File: code/mob/living/carbon/human/procs/emote.dm [hunk 8]
```diff
@@ -334,32 +366,51 @@
...
-						if ("glare","stare","look")
+						if ("nodto")
+							message = "<B>[src]</B> nods to [param]."
+							maptext_out = "<I>nods to [param]</I>"
+						if ("glare","stare","look","leer")
 							message = "<B>[src]</B> [act]s at [param]."
 							maptext_out = "<I>[act]s at [param]</I>"
```
Comment (ZeWaka) [inline code/mob/living/carbon/human/procs/emote.dm:415]:
```suggestion
						if ("nodto")
							message = "<B>[src]</B> nods to [M]."
							maptext_out = "<I>nods to [M]</I>"
						if ("glare","stare","look","leer")
							message = "<B>[src]</B> [act]s at [M]."
							maptext_out = "<I>[act]s at [M]</I>"
```

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**